trio_websocket/__init__.py

```diff
@@ -153,6 +156,104 @@ def _url_to_host(url, ssl_context):
     return url.host, url.port, resource, ssl_context
 
 
+async def wrap_client_stream(nursery, stream, host, resource, *,
+    task_status=trio.TASK_STATUS_IGNORED):
+    '''
+    Wrap an arbitrary stream in a client-side ``WebSocketConnection``.
+
+    This is a low-level function only needed in rare cases. Most users should
+    call ``open_websocket()`` or ``open_websocket_url()``.
+
+    This function supports the Trio nursery protocol. It should be started like
+    this: ``conn = await nursery.start(wrap_client_stream, …)``. It will suspend
+    until the connection's WebSocket open handshake is complete and then return
+    the connection object.
+
+    :param nursery: A Trio nursery to run background tasks in.
+    :param stream: A Trio stream to be wrapped.
+    :param str host: A host string that will be sent in the ``Host:`` header.
+    :param str resource: A resource string, i.e. the path component to be
+        accessed on the server.
+    :param task_status: part of Trio nursery protocol
+    :rtype: WebSocketConnection
+    '''
+    wsproto = wsconnection.WSConnection(wsconnection.CLIENT, host=host,
+        resource=resource)
+    connection = WebSocketConnection(stream, wsproto, path=resource)
+    nursery.start_soon(connection._reader_task)
+    await connection._open_handshake.wait()
+    task_status.started(connection)
```

Comment:
This doesn't look right.  reader_task should be the function supporting start, and should do the open_handshake wait, so that the block happens in the user's nursery.

(There's no reason to use started when you're returning on the next line.)

Reply:
Sorry, actually start protocol isn't needed anywhere.  Just await on open_handshake as you are now and return the connection.  Caller just does `connection = await serve_websocket(...)`